### DISCUSSION TITLE
Which hardware are you using to run your CI

### DISCUSSION THREAD

**Discussion by matiasvillaverde** [MobileNativeFoundation/discussions#23] [General]
Hi Folks, I start this thread to talk about which hardware are you using to run your CI.

Do you run it in-house?
Which machines are you using? 
Which tools do you find useful for orchestration?  
Which pain points do you have?
Do you have a dedicated team working on CI?

**Comment by BalestraPatrick** [comment]
I can talk a bit about what we do on the iOS side at Spotify since we wrote extensively about our macOS CI experience [here](https://engineering.atspotify.com/2020/05/01/how-we-gave-superpowers-to-our-macos-ci/).

We use [Flow](https://flow.swiss) to power all of our macOS CI. It's a great service that takes care of providing us Baremetal machines (we encountered a lot of performance issues with virtual machines). We mostly use Mac Mini 2018 with 32GB of RAM and 512GB SSD (you can quickly run out of disk space when installing multiple versions of the same developers tools (Xcode, Android Studio, different SDKs in one agent). We find out that 16GB of RAM were not enough for us and the upgrade to the 32GB of RAM tier was worth the price. I recommend to spend some time benchmarking different machine configurations to find what suit best your project (difference dependency graphs may benefit from different configurations for example).

We came up with a set of benchmark jobs that would exercise common CI tasks and run them on different configurations. We were able to quickly test new machines pretty easily in this way (M1 and Mac Pro for example).

We use Puppet to orchestrate the machines and roll out changes to them. We can run into some problems regarding our network when rolling out changes to hundreds of machines at once, so sometimes we have to create some custom rollouts that make sure that machine don't pull the latest configuration all at once.

We have a dedicated team working on CI infrastructure for our client builds, as well as other teams in the company that take care of the CI infrastructure for other disciplines. They collaborate and share a lot of tools.

**Reply by raven** [reply]
You mentioned "Android Studio", do you run some Android build tasks on the Mac minis also? Just curious about the Android tooling.

**Comment by dalemyers** [comment]
For the iOS side for Outlook we simply used the hosted agents that Azure DevOps provides: https://docs.microsoft.com/en-us/azure/devops/pipelines/agents/hosted?view=azure-devops&tabs=yaml 

This means that we don't have to worry about installing software, updating images, etc. It makes our day to day lives significantly easier. The down side though is that we don't control the update schedule for software. The images get updated once a week. If you are waiting on a beta version of Xcode or whatever, you may end up having to wait the full week depending on when it is released. We've found that overall, it's worth it though. If there's something we _really_ need, we can always adjust our tooling to install what we need. 

As for a CI "team", we don't really have that. I work full time on what we call "Developer Agility". That covers anything and everything which helps us ship better code faster. Most of that is our CI system, but it covers other things like ensuring that we are using the right tooling locally (including creating our own), looking at project structure, dependencies, etc. to see how these could be optimised, working on guides to help new developers get up and running, etc. Our agility team consists of myself, and one other person. They focus more on the build speed and code quality aspects (with a move to Bazel being their current big project).

**Comment by keith** [comment]
At Lyft we self host 100 mac minis in our own data centers. We have the 32gb intel models, hardware has been good so far except I kind of wish we had go for more disk space since some of our larger test builds can run out of the 250gb of storage we have. When we made the decision to move this in house the offerings out there for macOS CI weren't very flexible and were still quite expensive. We really wanted the bare metal performance as well as to be able to upgrade Xcode versions soon after release (and test betas).

We run buildkite directly on them without VMs. So far from what we've heard from folks the VM overhead just isn't worth it to us as long as we can make sure we don't pollute the environment. This has worked well for us since our CI and tooling is maintained by our "client tooling" team, so we don't have issues with other folks changing CI scripts or anything very often.

We are currently still using the easy path of managing them with [Apple Remote Desktop](https://www.apple.com/remotedesktop/index.html). We haven't had any issues that required completely re-imaging machines or anything.

We also maintain a k8s cluster of linux machines for Android CI and the iOS jobs wherever we can not run on macOS. We use these with buildkite as well and unsurprisingly that works quite well with very little maintenance.

For us the biggest pain point on the iOS side at this point is test flakiness related to the iOS simulator. I think we've made some headway on it recently but in the past it seems like OS contention lead to tests failing quite often in ways developers could never reproduce.

**Reply by mattrobmattrob** [reply]
>We also maintain a k8s cluster of linux machines for Android CI and the iOS jobs wherever we can not run on macOS.

This is one reason I was so stoked about https://github.com/yonaskolb/XcodeGen/pull/988 enabling Linux usage of `xcodegen`. So much faster and cheaper for some tests that we have written to lint Xcode project structure/other things. 👏 And generally pretty easy to run the latest Swift linux Docker image with a given SPM dep if it's Linux compatible.

**Reply by keith** [reply]
Awesome! I hadn't seen that this had landed. I don't think we currently have a specific use case for running it there, but I'm glad we can now since I also wanted this in the past

**Reply by jshchnz** [reply]
Have you considered switching to Amazon’s new EC2 Mac Instances? I know cost has been a big reason not to move over, curious if it was considered

**Reply by keith** [reply]
Yea we looked at it. I would love to be out of the business of hosting the machines but it definitely seemed cost prohibitive so we're not investigating it.

**Reply by beefon** [reply]
> For us the biggest pain point on the iOS side at this point is test flakiness related to the iOS simulator.

Ah, we have a couple of hacks to make this reliable. 

The most important hack is called "Cleanup and Reboot Nightly". You can guess what it does :) Without it `CoreSimulatorService` may just start freezing randomly, e.g. even on `simctl list` calls.

The second hack is to manage simulators lifecycle - [Emcee](http://github.com/avito-tech/Emcee/) manages the simulator farm and shutdowns and even deletes unneeded (idle) simulators, or when the number of working simulators exceeds a maximum count for a given machine. But we tend to re-use simulators if possible, i.e. when they are booted and iOS runtime matches one from a test run.

Also, we try to boot simulators multiple times in case if initial boot attempt fails.

**Reply by raven** [reply]
>  as long as we can make sure we don't pollute the environment

@keith can you expand on any tools / practices you use to achieve the clean environment?

The primary reason we have our VM setup is to ensure build reproducibility, as we have in the past been bitten by Simulator Service instability growing significantly with time (regardless of which services we restart). Just as you said, OS <-> iOS Simulator is a very fragile environment to manage reliably. We still see flakiness with UI test suites even with the VM setup we have. 

>  I think we've made some headway on it recently but in the past it seems like OS contention lead to tests failing quite often

Details on this would be ✨

**Reply by keith** [reply]
> @keith can you expand on any tools / practices you use to achieve the clean environment?

we're not doing anything super technically smart here. We mostly only run commands through bazel which handles everything inside of it, outside of that we're just thoughtful with our CI scripts (which our team owns all of) to not pollute global state. Then we have a bunch of cleanup logic for various issues like the simulator ones you mentioned (again nothing crazy but we delete them all between jobs)

>I think we've made some headway on it recently but in the past it seems like OS contention lead to tests failing quite often

It seems like at the very least in our case since we were building and running tests in parallel with bazel that we were oversubscribing the CPU. We decreased the parallelism there to help. I've also heard others do their build and test separately but still test in parallel, but I think for us we have some benefit from starting testing sooner since that's our critical path.

**Reply by RufusMall** [reply]
@keith In regards to your flaky tests on the simulator - are you using: https://github.com/pointfreeco/swift-snapshot-testing to do your snapshot testing? And are these snapshot tests interleaved with your unit tests?

I recently discovered issues with that framework, where UIKit objects/CAAnimation transactions/cleanup operations are being scheduled to execute on the next runloop - so they execute outside the lifecycle of the snapshot test. This means if a unit test calls some asynchronous code and uses XCTestExpectation to wait for the code to complete, it would struggle to get back to the main thread because the next cycle of the run loop would try to finalise the CATransactions and then release those objects - blocking the main thread for a significant amount of time and causing the unit test to time out.

I wonder if you are/were seeing a similar issue here.

**Comment by iosdevben** [comment]
At carsales, two of our app development teams share an in-house CI setup that is comprised of an iMac running TeamCity server and an agent, and two Mac Minis for agents. The whole setup is maintained by all the developers in the iOS app development teams.

These two teams are working on separate apps, and Fastlane powers our builds, but we don't have a centralised CI module. It was basically a copy-paste then customised to suit each team.

We built a "mock" server with Kitura to serve our UI tests, and this server runs on one of the Mac Minis.

Our biggest pain point is the UI tests: we've built lots of helper functions to try to make the tests more stable, but it's not stable enough that we can use the tests as a mandatory prerequisite for release builds.

**Comment by donnie-jp** [comment]
For those that do it in-house, what are the reasons you can't use a full cloud solution (for example bitrise)?
Running costs? Security of code? Something else?

Last year we moved all our iOS CI to bitrise from an in-house Jenkins setup and I'm very happy we did that!

**Reply by keith** [reply]
I think some of this has improved since we moved in house, but when we did all the offerings we evaluated were very expensive, very inflexible wrt Xcode updates and caching between builds, and slow with VM performance. So overall we felt like the benefits of those were worth the downsides of hosting the machines.

**Reply by iosdevben** [reply]
For us it's cost.

**Reply by shepting** [reply]
Yeah, we are pretty happy as well since places like MacStadium have a team of technicians to do manual work rather than paying your higher-paid engineers to do that.

**Comment by JoeSSS** [comment]
At XING we host MacMinis and MacPros in our DC and run VMs using Anka inside these MacMinis and MacPros. We are quite happy about the setup and its maintainability. The main advantage of Anka for us is creating templates and have them versioned by Xcode version, which can spin a new VM with the needed Xcode version in a matter of seconds and start building/testing immediately. We measured around 15% of performance decrease because of VM overhead in the past, but for us, it worth it because of better CI capacities with not that many MacMinis in the DC involved + orchestrating the VMs is very easy and clean.

**Comment by shepting** [comment]
At Airbnb we have 250 Mac mini CI machines at MacStadium (32GB, 6 core, 3.2GHz). They will handle wiping the machines occasionally if we need it for various reasons. We provision the machines mostly with Puppet, and then run Buildkite as our CI tooling.

We have a "Native Tools" team of 3 iOS folks to manage our infrastructure regarding CI and and Buck as our build system.

**Reply by mattrobmattrob** [reply]
With the Puppet + Buildkite setup, what part of your system does the "clean up" style tasks after a given build/task/test/what-unit-of-build has finished? The Buildkite agent functionality, Buck test host/runner, perhaps your own code? I'm thinking things like closing simulators, restarting macOS hosts, clearing derived data (in a non-Buck world), and anything else you all need to do!

**Comment by alexs555** [comment]
at Avito (iOS),

We use a farm of 75 Mac minis that we self-host and maintain by ourselves in our data center. 

We continuously enlarge our farm, so we have both devices of 2014 as well as new Mac minis from 2018, and a couple of M1 too. With specs from 16 to 32 GB RAM and from 128 to 512 GB SSD.

We use Puppet to configure them and to roll out updates. 

To check whether the Mac is ready for usage we run a special “build machine readiness check” script that runs all our everyday tasks on the new Mac. If this check succeeds, Mac mini becomes a Teamcity agent.

In comparison with MacStadium prices self-hosted solution was cheaper. iOS infra team consists of 4 devs who maintain the farm, develop testing tools and CI infrastructure.

**Comment by raven** [comment]
At realestate.com.au we have (iOS)

**Hardware**: Mac minis in MacStadium (2018 32GB RAM 512GB SSD)
**Hardware management/provisioning**: Ansible
**Virtualisation**: Anka with template VM images per Xcode version, each suspended from a 'warm' state with Simulators pre-launched.
**CI Platform**: Buildkite

**How it hangs together**:
Each metal machine has three Buildkite agents running on them with varying priorities for uniform distribution of builds across the metal. When a Buildkite build comes in, a custom bootstrap script grabs the desired toolchain from the build pipeline which is used to choose the Anka VM template that is then cloned, launched and has some just-in-time env/secrets applied. From there the Buildkite job is handed to the internal-to-vm Buildkite agent that runs job. Cleanup is just deleting the VM image once the job completes. 
Challenges we've had is the 512GB SSDs don't hold the greatest collections of VMs, gotta love consumer hardware being the backbone of enterprise solutions 😄 

**Why this setup**:
The entire organisation has heavily invested in Buildkite and has built great tooling it around it, so we wanted to align the iOS teams as closely as possible to reap the benefits. Virtualisation was chosen as we wanted to ensure repeatable ephemeral builds, as well as evaluating the impacts of concurrent builds on the same metal. Because we have multiple apps from multiple teams that have varied toolchain histories and practices, we have multiple VM templates with different toolchain versions, this is so building a new image from scratch doesn't take as long.

**Reply by hokadiri** [reply]
On average How long does it take from a when a  buildkite build come in to when a build kit agent runs the job?

**Comment by hokadiri** [comment]
Most master-agent CI setups like jenkins are active-passive. Curios if anyone has an active-active master-agent architecture and would be open to share how its setup.
